mm2src/coins/utxo/rpc_clients.rs

```diff
@@ -2388,32 +2546,445 @@ impl UtxoRpcClientOps for ElectrumClient {
+        debug!(
+            "Duplicate suspend timeout for address: {} from: {} to: {}",
+            address, suspend_timeout, new_timeout
+        );
+        *suspend_timeout = new_timeout;
+        Ok(())
+    }
+
+    async fn reset_suspend_timeout(state: &mut MutexGuard<'_, ConnMngState>, address: &str) -> Result<(), String> {
+        let suspend_timeout = &mut state
+            .conn_ctxs
+            .get_mut(address)
+            .ok_or_else(|| ERRL!("Failed to duplicate suspend_timeout for address: {}, no entry", address))?
+            .suspend_timeout_sec;
+        debug!(
+            "Reset supsend timeout for address: {} from: {} to the initial value: {}",
+            address, suspend_timeout, SUSPEND_TIMEOUT_INIT_SEC
+        );
+        *suspend_timeout = SUSPEND_TIMEOUT_INIT_SEC;
+        Ok(())
+    }
+
+    async fn restore_server(self, addr: String) {
+        debug!("Put back the connection to addr: {} after timeout: {}", addr, 5);
+        let mut guard = self.0.guarded.lock().await;
+        match guard
+            .conn_ctxs
+            .get(&addr)
+            .as_ref()
+            .expect("TODO") // TODO: get rid of unwrap
+            .conn_settings
+            .priority
+        {
+            Priority::Primary => guard.queue.primary.push_back(addr),
+            Priority::Secondary => guard.queue.backup.push_back(addr),
+        }
+
+        if let Some(current_addr) = guard.active.as_ref() {
+            debug!(
+                "Electrum: {} - is currently active, may be it should be substituted with the primary one",
+                current_addr
+            );
+        } else {
+            drop(guard);
+            let _ = self.connect().await;
+        };
+    }
+}
+
+#[derive(Debug)]
+struct ConnMngQueue {
+    primary: LinkedList<String>,
+    backup: LinkedList<String>,
+}
+
+#[async_trait]
+impl ConnMngTrait for ConnMng {
+    async fn get_conn(&self, address: Option<&str>) -> Option<Arc<AsyncMutex<ElectrumConnection>>> {
+        debug!("Getting connection for address: {:?}", address);
+        let guard = self.0.guarded.lock().await;
+
+        let address: String = {
+            if address.is_some() {
+                address.map(|internal| internal.to_string())
+            } else {
+                guard.active.as_ref().cloned()
+            }
+        }?;
+
+        guard.conn_ctxs.get(&address).map(|ctx| ctx.connection.clone())?
+    }
+
+    async fn connect(&self) -> Result<(), String> {
+        let weak_spawner = self.0.abortable_system.weak_spawner();
+
+        struct ConnectingStateCtx(ConnMng, WeakSpawner);
+        impl Drop for ConnectingStateCtx {
+            fn drop(&mut self) {
+                let spawner = self.1.clone();
+                let conn_mng = self.0.clone();
+                spawner.spawn(async move {
+                    let state = conn_mng.0.guarded.lock().await;
+                    state.connecting.store(false, AtomicOrdering::Relaxed);
+                })
```

Comment:
Can you please elaborate on this more? `ConnectingStateCtx` will be considered dropped before connecting is set to false if the lock was held by other operation for sometime.

Reply:
Certainly should have been more detailed! Thanks for pointing this out! This place is the heart of selective connectivity.

[done](https://github.com/KomodoPlatform/komodo-defi-framework/pull/1966/commits/9031c9f1ea2e4030510bff1c33a7d749e1b79d83)